### PR TITLE
New charm metadata

### PR DESCRIPTION
**Pull request by ycliuhw**
## Description of change

Various changes to get new charms to work with k8s.

This PR is based on https://github.com/juju/juju/pull/12051

- did some fixes;
- added proper tests;
- added a TODO: 
```
The volumeMount.MountPath was defined in `caas.ApplicationConfig.Filesystems[*].Attachment.Path`.
Consolidate `caas.ApplicationConfig.Filesystems[*].Attachment.Path` and `caas.ApplicationConfig.Containers[*].Mounts[*].Path`
```

### ComputedSeries

ComputedSeries is a newly introduced charm metadata field for v2 charms that use systems.
Since there is a large amount of logic in Juju that depends on series as a string, ComputedSeries returns systems as either their canonical series name like "focal" or as a complex string "system#resource=charm-resource-name".

Systems as a series string can support resource images or they can support simple os and unknown versions like "system#os=ubuntu#version=21.04"

### Requirements

**Needs unit tests**

Depends on:
- https://github.com/juju/charm-base-images/pull/2

## QA steps

```console
$ juju add-model t1 microk8s

$ juju deploy ./cockroach --resource cockroachdb-image=cockroachdb/cockroach:v20.1.4

$ kubectl get all,ing,pv,pvc -o wide -n t1
NAME                                READY   STATUS    RESTARTS   AGE   IP            NODE     NOMINATED NODE   READINESS GATES
pod/cockroachdb-0                   2/2     Running   0          19m   10.1.78.110   ubuntu   <none>           <none>
pod/modeloperator-78c8c4b4f-8bgcw   1/1     Running   0          24m   10.1.78.109   ubuntu   <none>           <none>

NAME                            TYPE        CLUSTER-IP       EXTERNAL-IP   PORT(S)     AGE   SELECTOR
service/cockroachdb             ClusterIP   10.152.183.195   <none>        65535/TCP   19m   juju-app=cockroachdb
service/cockroachdb-endpoints   ClusterIP   None             <none>        <none>      19m   juju-app=cockroachdb
service/modeloperator           ClusterIP   10.152.183.177   <none>        17071/TCP   24m   juju-modeloperator=modeloperator

NAME                            READY   UP-TO-DATE   AVAILABLE   AGE   CONTAINERS      IMAGES                             SELECTOR
deployment.apps/modeloperator   1/1     1            1           24m   juju-operator   ycliuhw/jujud-operator:2.9-beta1   juju-modeloperator=modeloperator

NAME                                      DESIRED   CURRENT   READY   AGE   CONTAINERS      IMAGES                             SELECTOR
replicaset.apps/modeloperator-78c8c4b4f   1         1         1       24m   juju-operator   ycliuhw/jujud-operator:2.9-beta1   juju-modeloperator=modeloperator,pod-template-hash=78c8c4b4f

NAME                           READY   AGE   CONTAINERS                    IMAGES
statefulset.apps/cockroachdb   1/1     19m   juju-unit-agent,cockroachdb   ycliuhw/ubuntu:20.04,cockroachdb/cockroach:v20.1.4

NAME                                                        CAPACITY   ACCESS MODES   RECLAIM POLICY   STATUS   CLAIM                                            STORAGECLASS        REASON   AGE   VOLUMEMODE
persistentvolume/pvc-0797c052-baaa-4ef8-8858-860b38205274   20Gi       RWO            Delete           Bound    controller-k1/storage-controller-0               microk8s-hostpath            31m   Filesystem
persistentvolume/pvc-58ab2fd3-15a9-40eb-8125-3b8478589ca4   1Gi        RWO            Delete           Bound    t1/cockroachdb-database-e94caa18-cockroachdb-0   microk8s-hostpath            19m   Filesystem

NAME                                                                STATUS   VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS        AGE   VOLUMEMODE
persistentvolumeclaim/cockroachdb-database-e94caa18-cockroachdb-0   Bound    pvc-58ab2fd3-15a9-40eb-8125-3b8478589ca4   1Gi        RWO            microk8s-hostpath   19m   Filesystem

$ kubectl get pod/cockroachdb-0 -n t1 -o json | jq '{containers: .spec.containers, initContainers: .spec.initContainers}'
{
  "containers": [
    {
      "args": [
        "unit",
        "--data-dir",
        "/var/lib/juju"
      ],
      "command": [
        "/usr/bin/k8sagent"
      ],
      "env": [
        {
          "name": "JUJU_CONTAINER_NAMES",
          "value": "cockroachdb"
        }
      ],
      "image": "ycliuhw/ubuntu:20.04",
      "imagePullPolicy": "IfNotPresent",
      "name": "juju-unit-agent",
      "resources": {},
      "terminationMessagePath": "/dev/termination-log",
      "terminationMessagePolicy": "File",
      "volumeMounts": [
        {
          "mountPath": "/usr/bin/k8sagent",
          "name": "juju-data-dir",
          "readOnly": true,
          "subPath": "usr/bin/k8sagent"
        },
        {
          "mountPath": "/var/lib/juju",
          "name": "juju-data-dir",
          "subPath": "var/lib/juju"
        },
        {
          "mountPath": "/var/run/containers",
          "name": "juju-data-dir",
          "subPath": "var/run/containers"
        },
        {
          "mountPath": "/var/lib/juju/storage/database/0",
          "name": "cockroachdb-database-e94caa18"
        }
      ],
      "workingDir": "/var/lib/juju"
    },
    {
      "command": [
        "/usr/bin/pebble"
      ],
      "env": [
        {
          "name": "JUJU_CONTAINER_NAME",
          "value": "cockroachdb"
        }
      ],
      "image": "cockroachdb/cockroach:v20.1.4",
      "imagePullPolicy": "IfNotPresent",
      "name": "cockroachdb",
      "resources": {},
      "terminationMessagePath": "/dev/termination-log",
      "terminationMessagePolicy": "File",
      "volumeMounts": [
        {
          "mountPath": "/usr/bin/pebble",
          "name": "juju-data-dir",
          "readOnly": true,
          "subPath": "usr/bin/pebble"
        },
        {
          "mountPath": "/var/run/container",
          "name": "juju-data-dir",
          "subPath": "var/run/containers/cockroachdb"
        },
        {
          "mountPath": "/cockroach/cockroach-data",
          "name": "cockroachdb-database-e94caa18"
        }
      ]
    }
  ],
  "initContainers": [
    {
      "args": [
        "init"
      ],
      "command": [
        "/opt/k8sagent"
      ],
      "env": [
        {
          "name": "JUJU_CONTAINER_NAMES",
          "value": "cockroachdb"
        },
        {
          "name": "JUJU_K8S_POD_NAME",
          "valueFrom": {
            "fieldRef": {
              "apiVersion": "v1",
              "fieldPath": "metadata.name"
            }
          }
        },
        {
          "name": "JUJU_K8S_POD_UUID",
          "valueFrom": {
            "fieldRef": {
              "apiVersion": "v1",
              "fieldPath": "metadata.uid"
            }
          }
        }
      ],
      "envFrom": [
        {
          "secretRef": {
            "name": "cockroachdb-application-config"
          }
        }
      ],
      "image": "ycliuhw/k8sagent:2.9-beta1",
      "imagePullPolicy": "IfNotPresent",
      "name": "juju-unit-init",
      "resources": {},
      "terminationMessagePath": "/dev/termination-log",
      "terminationMessagePolicy": "File",
      "volumeMounts": [
        {
          "mountPath": "/var/lib/juju",
          "name": "juju-data-dir",
          "subPath": "var/lib/juju"
        },
        {
          "mountPath": "/shared/usr/bin",
          "name": "juju-data-dir",
          "subPath": "usr/bin"
        },
        {
          "mountPath": "/var/run/containers",
          "name": "juju-data-dir",
          "subPath": "var/run/containers"
        }
      ],
      "workingDir": "/var/lib/juju"
    }
  ]
}
```

## Documentation changes

Many changes need to be documented around charm metadata, environment variables, shared container mount paths etc.

## Bug reference

N/A

### CODE REVIEW

**Comment by ycliuhw** [comment]
[Upgrade charm/v8 to support the metadata v2 in juju](https://trello.com/c/hhYLWQmK/1199-upgrade-charm-v8-to-support-the-metadata-v2-in-juju)

File: apiserver/facades/controller/caasapplicationprovisioner/provisioner.go [hunk 1]
```diff
@@ -4,14 +4,18 @@
...
 	"sort"
 
-	"github.com/juju/charm/v8"
+	charmresource "github.com/juju/charm/v8/resource"
+	"github.com/juju/collections/set"
 	"github.com/juju/errors"
 	"github.com/juju/loggo"
 	"github.com/juju/names/v4"
-	"github.com/juju/utils/set"
+	"gopkg.in/yaml.v3"
```
Comment (wallyworld) [inline apiserver/facades/controller/caasapplicationprovisioner/provisioner.go:18]: Juju codebase currently uses yaml.v2

File: apiserver/facades/controller/caasapplicationprovisioner/provisioner.go [hunk 2]
```diff
@@ -327,15 +319,12 @@ func (a *API) garbageCollectOneApplication(args params.CAASApplicationGarbageCol
...
 	if err != nil {
 		return errors.Trace(err)
-	}
-	if ch.Meta() == nil ||
-		ch.Meta().Deployment == nil {
-		return errors.Errorf("charm missing deployment info")
-	}
-	deploymentType := ch.Meta().Deployment.DeploymentType
+	}*/
+	// TODO(new-charms): support more than statefulset
```
Comment (wallyworld) [inline apiserver/facades/controller/caasapplicationprovisioner/provisioner.go:326]: move above comment

File: apiserver/facades/controller/caasapplicationprovisioner/provisioner.go [hunk 3]
```diff
@@ -310,13 +302,13 @@ func (a *API) garbageCollectOneApplication(args params.CAASApplicationGarbageCol
 	if err != nil {
 		return errors.Trace(err)
 	}
-	observedUnitTags := set.NewTags()
+	observedUnitTags := set.NewStrings()
```
Comment (wallyworld) [inline apiserver/facades/controller/caasapplicationprovisioner/provisioner.go:305]: Why change from tags to strings?

File: apiserver/facades/controller/caasapplicationprovisioner/provisioner.go [hunk 4]
```diff
@@ -615,3 +604,77 @@ func (a *API) devicesParams(app Application) ([]params.KubernetesDeviceParams, e
...
+		}
+		if res.Results[i].Error != nil {
+			continue
+		}
+		res.Results[i].Result = &imageResources
+	}
+	return res, nil
+}
+
+func readDockImageResource(reader io.Reader) (params.DockerImageInfo, error) {
```
Comment (wallyworld) [inline apiserver/facades/controller/caasapplicationprovisioner/provisioner.go:661]: readDockerImageResource

File: caas/kubernetes/provider/application/application.go [hunk 5]
```diff
@@ -424,6 +422,10 @@ func (a *app) configureDefaultService(annotation annotations.Annotation) (err er
 		Spec: corev1.ServiceSpec{
 			Selector: a.labels(),
 			Type:     corev1.ServiceTypeClusterIP,
+			Ports: []corev1.ServicePort{{
```
Comment (wallyworld) [inline caas/kubernetes/provider/application/application.go:425]: Can we add a TODO

File: caas/kubernetes/provider/precheck.go [hunk 6]
```diff
@@ -39,9 +39,10 @@ func (k *kubernetesClient) PrecheckInstance(ctx context.ProviderCallContext, par
 		return errors.NotSupportedf("constraints %v", strings.Join(unsupported, ","))
 	}
 
-	if params.Series != k8sSeries {
-		return errors.NotValidf("series %q", params.Series)
-	}
+	// TODO(new-charms): handle systems
```
Comment (wallyworld) [inline caas/kubernetes/provider/precheck.go:42]: TODO(embedded)

for consistency

Can we still check legacy charms though?
We can see if params.Series == k8sSeries or matches the patterm for embedded charms

File: caas/kubernetes/provider/precheck_test.go [hunk 7]
```diff
@@ -37,6 +37,8 @@ func (s *PrecheckSuite) TestSuccess(c *gc.C) {
 }
 
 func (s *PrecheckSuite) TestWrongSeries(c *gc.C) {
+	c.Skip("disable for now because TODO(new-charms): handle systems")
```
Comment (wallyworld) [inline caas/kubernetes/provider/precheck_test.go:40]: I think we should support matching on "kubernetes" or checking the pattern

File: core/model/model.go [hunk 8]
```diff
@@ -41,19 +42,35 @@ type Model struct {
...
-	os, err := series.GetOSFromSeries(charmSeries)
-	if err != nil {
-		return errors.Trace(err)
-	}
-	switch modelType {
-	case CAAS:
-		if !caasOS.Contains(os.String()) {
-			return errors.NotValidf("series %q in a kubernetes model", charmSeries)
+func ValidateSeries(modelType ModelType, charmSeries string, charmV2 bool) error {
+	if charmV2 {
```
Comment (wallyworld) [inline core/model/model.go:46]: maybe pass in charm version as an int and use switch to allow for v3 etc

File: core/model/model_test.go [hunk 9]
```diff
@@ -29,7 +29,27 @@ func (*ModelSuite) TestValidateSeries(c *gc.C) {
...
+
+func (*ModelSuite) TestValidateSeriesNewCharm(c *gc.C) {
+	for _, t := range []struct {
+		modelType model.ModelType
+		series    string
+		valid     bool
+	}{
+		{model.IAAS, "bionic", true},
+		{model.IAAS, "bionic", true},
+		{model.CAAS, "bionic", true},
```
Comment (wallyworld) [inline core/model/model_test.go:49]: also include new examples with system etc

File: state/charm.go [hunk 10]
```diff
@@ -724,7 +724,7 @@ func validateCharmSeries(modelType ModelType, series string, ch hasMeta) error {
 	if series == "" {
 		return nil
 	}
-	return model.ValidateSeries(model.ModelType(modelType), series)
+	return model.ValidateSeries(model.ModelType(modelType), series, ch.Meta().Format() >= charm.FormatV2)
```
Comment (wallyworld) [inline state/charm.go:727]: Just pass in the format

File: testcharms/charm-repo/focal/cockroach/revision [hunk 11]
```diff
@@ -0,0 +1 @@
+0
```
Comment (wallyworld) [inline testcharms/charm-repo/focal/cockroach/revision:1]: new line?

File: worker/caasapplicationprovisioner/worker_test.go [hunk 12]
```diff
@@ -132,7 +131,7 @@ func (s *CAASApplicationSuite) TestWorkerStartOnceNotify(c *gc.C) {
 
 	c.Assert(called, gc.Equals, 1)
 	c.Assert(notifyWorker, gc.NotNil)
+	time.Sleep(coretesting.ShortWait)
```
Comment (wallyworld) [inline worker/caasapplicationprovisioner/worker_test.go:134]: Don't use sleeps, we need to use channels instead

**Review by wallyworld — APPROVED** [review]
Thanks for picking this up. Several small things, nothing too major.

File: apiserver/common/applicationwatcher.go [hunk 13]
```diff
@@ -163,7 +163,7 @@ func (w *applicationWatcher) handle(changes []string) ([]string, error) {
 }
 
 // Changes is part of corewatcher.StringsWatcher.
-func (w *applicationWatcher) Changes() corewatcher.StringsChannel {
+func (w *applicationWatcher) Changes() <-chan []string {
```
[the remark this answers is not in the document]
Comment (ycliuhw) [reply apiserver/common/applicationwatcher.go:166]: there is somewhere a type error, so made this change

File: apiserver/facades/controller/caasapplicationprovisioner/provisioner.go (hunk 3 above)
Comment (ycliuhw) [reply apiserver/facades/controller/caasapplicationprovisioner/provisioner.go:305]: I don't know, foundUnits changed as well

File: caas/kubernetes/provider/application/application.go (hunk 5 above)
Comment (ycliuhw) [reply caas/kubernetes/provider/application/application.go:425]: not sure for what?

File: caas/kubernetes/provider/precheck.go (hunk 6 above)
Comment (ycliuhw) [reply caas/kubernetes/provider/precheck.go:42]: We still need do this kinda check, u see there is a TODO for this.
```go
if params.Series == k8sSeries && meta.Format()>=2{return notValidErr}
```

File: caas/kubernetes/provider/precheck_test.go (hunk 7 above)
Comment (ycliuhw) [reply caas/kubernetes/provider/precheck_test.go:40]: It can be enabled after we remove the above todo

File: core/model/model_test.go (hunk 9 above)
Comment (ycliuhw) [reply core/model/model_test.go:49]: the `series` here is the result parsed from `systems`. 
So the `systems` is not accessible from here(`ValidateSeries()`) anymore.

File: caas/kubernetes/provider/precheck.go (hunk 6 above)
Comment (wallyworld) [reply caas/kubernetes/provider/precheck.go:42]: I mean for v1 charms. We are removing the check for those and so could have issues.
And even for v2, we can check the syntax of the series string

File: apiserver/common/applicationwatcher.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (wallyworld) [reply apiserver/common/applicationwatcher.go:166]: How is that?
We use corewatcher.StringsChannel as the return type on the interface method - it is the correct thing to do here to use it

[the remark this answers is not in the document]
Comment (wallyworld) [reply apiserver/common/applicationwatcher.go:166]: ah, i see we use []string in apiserver layer

File: caas/kubernetes/provider/precheck.go (hunk 6 above)
Comment (wallyworld) [reply caas/kubernetes/provider/precheck.go:42]: We discussed and this is ok

File: apiserver/facades/controller/caasapplicationprovisioner/provisioner.go (hunk 3 above)
Comment (wallyworld) [reply apiserver/facades/controller/caasapplicationprovisioner/provisioner.go:305]: @hpidcock do you know why this was changed?

**Pull request merged**